### PR TITLE
CI Make torch compile tests run on GPU

### PR DESCRIPTION
**Pull request by BenjaminBossan**
Many of these tests require a GPU to run, so use custom runners.

Code was mostly copied from existing workflows, I hope that this works.

### CODE REVIEW

File: .github/workflows/torch_compile_tests.yml [hunk 1]
```diff
@@ -11,32 +11,46 @@ on:
...
+env:
+  RUN_SLOW: "yes"
+  IS_GITHUB_CI: "1"
+  # To be able to run tests on CUDA 12.2
+  NVIDIA_DISABLE_REQUIRE: "1"
+
 jobs:
   run_tests_with_compile:
-    runs-on: ubuntu-latest
+    runs-on: [self-hosted, single-gpu, nvidia-gpu, t4, ci]
```
Comment (younesbelkada) [inline .github/workflows/torch_compile_tests.yml:22]:
```suggestion
    runs-on: [self-hosted, single-gpu, nvidia-gpu, a10, ci]
```

**Review by younesbelkada — APPROVED** [review]
Thanks ! Note we do have access to A10 GPUs in case you want to run them on more recent hardware

File: .github/workflows/torch_compile_tests.yml [hunk 2]
```diff
@@ -11,32 +11,46 @@ on:
...
+    strategy:
+      fail-fast: false
+      matrix:
+          docker-image-name: ["huggingface/peft-gpu-bnb-latest:latest"]
```
Comment (younesbelkada) [inline .github/workflows/torch_compile_tests.yml:26]: nit: we don't really need matrix here

Comment (younesbelkada) [inline .github/workflows/torch_compile_tests.yml:26]:
```suggestion
```

File: .github/workflows/torch_compile_tests.yml [hunk 3]
```diff
@@ -11,32 +11,46 @@ on:
...
     env:
       PEFT_DEBUG_WITH_TORCH_COMPILE: 1
+      CUDA_VISIBLE_DEVICES: "0"
+      TEST_TYPE: "single_gpu_${{ matrix.docker-image-name }}"
+    container:
+      image: ${{ matrix.docker-image-name }}
```
Comment (younesbelkada) [inline .github/workflows/torch_compile_tests.yml:32]:
```suggestion
      image: "huggingface/peft-gpu-bnb-latest:latest"
```

File: .github/workflows/torch_compile_tests.yml [hunk 4]
```diff
@@ -11,32 +11,46 @@ on:
...
-    runs-on: ubuntu-latest
+    runs-on: [self-hosted, single-gpu, nvidia-gpu, t4, ci]
+    strategy:
+      fail-fast: false
+      matrix:
+          docker-image-name: ["huggingface/peft-gpu-bnb-latest:latest"]
     env:
       PEFT_DEBUG_WITH_TORCH_COMPILE: 1
+      CUDA_VISIBLE_DEVICES: "0"
+      TEST_TYPE: "single_gpu_${{ matrix.docker-image-name }}"
```
Comment (younesbelkada) [inline .github/workflows/torch_compile_tests.yml:30]:
```suggestion
      TEST_TYPE: "single_gpu_huggingface/peft-gpu-bnb-latest:latest"
```

**Review by younesbelkada — COMMENTED** [review]
Left also small nits to make the workflow file clearer

**Comment by BenjaminBossan** [comment]
@younesbelkada Thanks for the review. I kept the test matrix with a single entry to make it easier to extend in the future but as is, it's indeed not needed, so it's removed now.

**Pull request merged**